pep-0646.rst

```diff
@@ -155,432 +95,344 @@ Type variable tuples are created with:
+was annotated as ``Tuple[*Shape]``. Why is this necessary - if ``Shape``
+behaves like ``Tuple[T1, T2, ...]``, couldn't we have annotated the ``shape``
+argument as ``Shape`` directly?
 
-::
+This is, in fact, deliberately not possible: type variable tuples must
+*always* be used unpacked (that is, prefixed by the star operator). This is
+for two reasons:
 
-    # NOT valid
-    def foo(**kwargs: Ts): ...
-    def foo(**kwargs: *Ts): ...
+* To avoid potential confusion about whether to use a type variable tuple
+  in a packed or unpacked form ("Hmm, should I do ``-> Shape``,
+  or ``-> Tuple[Shape]``, or ``-> Tuple[*Shape]``...?")
+* To improve readability: the star also functions as an explicit visual
+  indicator that the type variable tuple is not a normal type variable.
 
-Type Variable Tuples with ``Callable``
+``Unpack`` for Backwards Compatibility
 ''''''''''''''''''''''''''''''''''''''
 
-Type variable tuples can also be used in the arguments section of a
-``Callable``:
+Note that the use of the star operator in this context requires a grammar change,
+and is therefore only available in new versions of Python. To enable use of type
+variable tuples in older versions of Python, we introduce the ``Unpack`` type
+operator that can be used in place of the star operator to unpack type variable tuples:
 
 ::
 
-    class Process:
-      def __init__(target: Callable[[*Ts], Any], args: Tuple[*Ts]): ...
+  # Unpacking using the star operator in new versions of Python
+  class Array(Generic[*Shape]): ...
 
-    def func(arg1: int, arg2: str): ...
-    
-    Process(target=func, args=(0, 'foo'))  # Passes type-check
-    Process(target=func, args=('foo', 0))  # Fails type-check
-
-Type Variable Tuples with ``Union``
-'''''''''''''''''''''''''''''''''''
+  # Unpacking using ``Unpack`` in older versions of Python
+  class Array(Generic[Unpack[Shape]]): ...
 
-Finally, type variable tuples can be used with ``Union``:
+Variance, Type Constraints and Type Bounds: Not (Yet) Supported
+'''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
 
-::
-    
-    def f(*args: *Ts) -> Union[*Ts]:
-        return random.choice(args)
+To keep this PEP minimal, ``TypeVarTuple`` does not yet support specification of:
 
-    f(1, 'foo')  # Inferred type is Union[int, str]
+* Variance (e.g. ``TypeVar('T', covariant=True)``)
+* Type constraints (``TypeVar('T', int, float)``)
+* Type bounds (``TypeVar('T', bound=ParentClass)``)
 
-If the type variable tuple is empty (e.g. if we had ``*args: *Ts``
-and didn't pass any arguments), the type checker should
-raise an error on the ``Union`` (matching the behaviour of ``Union``
-at runtime, which requires at least one type argument).
+We leave the decision of how these arguments should behave to a future PEP, when variadic generics have been tested in the field. As of this PEP, type variable tuples are
+**invariant**.
 
-``Map``
--------
+Behaviour when Type Parameters are not Specified
```

Comment:
Ah, yeah, I decided to just standardise on British spelling throughout to avoid getting confused. (Likewise speling 'parameterise' with an s instead of a z.)